src/stan/model/indexing/deep_copy.hpp

```diff
@@ -39,9 +40,9 @@ inline const T& deep_copy(const T& x) {
  * @param a Input matrix, vector, or row vector.
  * @return Deep copy of input.
  */
-template <typename T, int R, int C>
-inline Eigen::Matrix<T, R, C> deep_copy(const Eigen::Matrix<T, R, C>& a) {
-  Eigen::Matrix<T, R, C> result(a);
+template <typename EigMat, typename = require_eigen_t<EigMat>>
+inline auto deep_copy(const EigMat& a) {
+  typename EigMat::PlainObject result(a);
   return result;
```

Comment:
I think it's better to leave this as is since the intent of the function is force a copy and in the eigen docs for eval they say

```
Notice that in the case of a plain matrix or vector (not an expression) this function just returns a const reference, in order to avoid a useless copy.
```

https://eigen.tuxfamily.org/dox/classEigen_1_1DenseBase.html#aa73e57a2f0f7cfcb4ad4d55ea0b6414b

Though the function returns by value I think as is it's more clear this is making a copy